src/lib/vault/btc-provider.js

```diff
@@ -0,0 +1,183 @@
+import { Electrum } from './clients/electrum/electrum-client';
+import { NetworkUtil } from './network';
+import { coinSelection } from './clients/coin-selection/coin-selection';
+import { getRedeemScript, satsToBtc, signatureValidator } from './clients/utils/bitcoin-utils';
+import { txhexToElectrumTransaction } from './clients/utils/electrum-utils';
+
+const blockchain = Blockchain.BTC;
+
+// signer requirements for multisig vault
+const threshold = 2;
+
+const SEQUENCE_RBF_ENABLED = 0xffffffff - 2;
+
+// Network fee always set to fast
+const BTCFEERATE = 1;
+
+export const { signTransaction } = WalletService;
+
+class BitcoinProvider {
+  constructor() {
+    this.network = NetworkUtil.getNetwork(blockchain);
+    this.instance = new Electrum(this.network);
+  }
+
+  createPsbt(inputs, outputs, descriptor) {
+    const psbt = new bitcoin.Psbt({ network: this.network });
+    const redeem = this.getRedeemScriptFromDescriptor(descriptor);
+
+    psbt.addInputs(
+      inputs.map(input => {
+        const etx = txhexToElectrumTransaction(input.hex);
+
+        return {
+          hash: input.txid,
+          index: input.vout,
+          sequence: SEQUENCE_RBF_ENABLED,
+          witnessScript: redeem.output,
+          witnessUtxo: {
+            script: Buffer.from(etx.vout[input.vout].scriptPubKey.hex, 'hex'),
+            value: input.value
+          }
+        };
+      })
+    );
+
+    psbt.addOutputs(
+      outputs.map(output => {
+        return {
+          address: output?.address ?? this.multisig.address,
+          value: output?.value
+        };
+      })
+    );
+
+    return psbt;
+  }
+
+  async createTransaction({ to, from, descriptor }) {
+    const utxos = await this.instance.getAddressUTXOs(from);
+
+    if (utxos.length === 0) {
+      throw new Error('InsufficientFunds');
+    }
+
+    // Always use fast fee rate
+    const feeRate = await this.instance.estimateFee(BTCFEERATE);
+
+    // This means we will send all the funds to the destination
+    const output = [{ address: to }];
+
+    try {
+      const result = coinSelection(utxos, output, Number(feeRate));
+      const { inputs, outputs } = result;
+
+      const psbt = this.createPsbt(inputs, outputs, descriptor);
+
+      return psbt.toHex();
+    } catch (error) {
+      throw new Error(error.message);
+    }
+  }
+
+  deriveAddress(publicKey) {
+    return bitcoin.payments.p2wpkh({ network: this.network, pubkey: Buffer.from(publicKey, 'hex') }).address;
+  }
+
+  deriveMultisigAddress(descriptor) {
+    const redeem = this.getRedeemScriptFromDescriptor(descriptor);
+    const { address } = bitcoin.payments.p2wsh({ network: this.network, redeem });
+
+    return address;
+  }
+
+  async getAddressBalance(address) {
+    const { confirmed, unconfirmed } = await this.instance.getAddressBalance(address);
+
+    if (unconfirmed < 0) {
+      throw new Error('TransactionPending');
+    }
+
+    if (confirmed <= 0) {
+      throw new Error('AccountNotFound');
+    }
```

Comment:
this is also "Insufficient Funds" for BTC

Reply:
👍 Changed